File: src/stack/c/dynamicStack.c
#include <stdio.h>
#include <stdlib.h>
#include "dynamicStack.h"

struct DynamicStack* createStack(){
    struct DynamicStack *stack = malloc(sizeof(struct DynamicStack));
    stack->head = NULL;

    return stack;
}

int push(struct DynamicStack *stack, int value){
    struct Node *node = malloc(sizeof(struct Node));
    node->value = value;
    node->next = stack->head;
    stack->head = node;

    return 0;
}

int pop(struct DynamicStack *stack){
    if (isEmpty(stack)){
        printf("Stack is empty!\n");
        return -1;
    }

    struct Node *node = stack->head;
    int value = node->value;
    stack->head = stack->head->next;

    free(node);

    return value;
}

int peek(struct DynamicStack *stack){
    if (isEmpty(stack)){
        printf("Stack is empty!\n");
        return -1;
    }

    return stack->head->value;
}

int print(struct DynamicStack *stack){
    if (isEmpty(stack)){
        printf("Stack is empty!\n");
        return -1;
    }

    struct Node *current = stack->head;
    while (current != NULL){
        printf("[ %d ]\n", current->value);
        current = current->next;
    }
    printf("\n");

    return 0;
}

int isEmpty(struct DynamicStack *stack){
    return stack->head == NULL;
}

File: src/dataStructures/stack/ts/dynamicStack.ts
import { INode } from "./types"
import { EmptyStackError } from "./utils"

export class Node<Type> implements INode<Type> {
    constructor(
        public value: Type,
        public next: INode<Type> | null
    ) {}
}

export class DynamicStack<Type> {
    private head: INode<Type> | null

    constructor() {
        this.head = null
    }

    public push(value: Type): void {
        if (this.head === null) {
            this.head = new Node(value, null)
        } else {
            this.head = new Node(value, this.head)
        }
    }

    public pop(): Type {
        if (this.head === null) {
            throw new EmptyStackError()
        }

        const value = this.head.value
        this.head = this.head.next

        return value
    }

    public peek(): Type {
        if (this.head === null) {
            throw new EmptyStackError()
        }
        return this.head.value
    }

    public isEmpty(): boolean {
        return this.head === null
    }

    public print(): void {
        if (this.isEmpty()) {
            throw new EmptyStackError()
        }

        let output = ""
        let current = this.head

        while (current !== null) {
            output += `[ ${current.value} ]\n`
            current = current.next
        }

        console.log(output)
    }
}
